### PR TITLE
[ORCA] Fill Agg argtypes as optimizer_enable_eageragg is on

### PR DESCRIPTION
**Pull request by charliettxx**
To Fix issue https://github.com/greenplum-db/gpdb/issues/16978, In CXformEagerAgg rule, we chould fill argtypes of CLogicalAgg, which used for filling Aggref in finla Plan.

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/gporca/libgpopt/src/xforms/CXformEagerAgg.cpp [hunk 1]
```diff
@@ -357,7 +361,7 @@ CXformEagerAgg::PopulateUpperProjectElement(
 	agg_mdid->AddRef();
 	CScalarAggFunc *upper_agg_func = CUtils::PopAggFunc(
 		mp, agg_mdid, agg_name, is_distinct, EaggfuncstageGlobal, true, nullptr,
-		EaggfunckindNormal, GPOS_NEW(mp) ULongPtrArray(mp), false);
+		EaggfunckindNormal, argtypes, false);
```
Comment (dgkimura) [inline src/backend/gporca/libgpopt/src/xforms/CXformEagerAgg.cpp:364]: Something feels funny about the memory ownership here.  `CScalarAggFunc` releases `m_argtypes` in its destructor.  But I don't see any `AddRefs()` or `GPOS_NEW`'d objects in the new code path.  Without doing that shouldn't that cause a double free?

Comment (charliettxx) [reply src/backend/gporca/libgpopt/src/xforms/CXformEagerAgg.cpp:366]: oh I ignore it, yes we should `AddRefs()` for argtypes, and I've updated it.

**Review by chrishajas — APPROVED** [review]
Nice find! I also think it'd be worth adding a test in gporca.sql with the test case in the issue

File: src/backend/gporca/libgpopt/src/xforms/CXformEagerAgg.cpp [hunk 2]
```diff
@@ -349,15 +353,17 @@ CXformEagerAgg::PopulateUpperProjectElement(
...
 	CWStringConst *agg_name, CColRef *lower_colref, CColRef *output_colref,
 	BOOL is_distinct,
+	ULongPtrArray *argtypes,
 	CExpression **
 		upper_proj_elem_expr  // output project element of the new lower aggregate
 )
 {
 	// create a new operator
 	agg_mdid->AddRef();
+	argtypes->AddRef();
```
Comment (dgkimura) [inline src/backend/gporca/libgpopt/src/xforms/CXformEagerAgg.cpp:363]: Nit - `arg_types` seems more consistent style with surrounding variable names both in this file and in the header file.

**Review by dgkimura — APPROVED** [review]

**Comment by charliettxx** [comment]
> Nice find! I also think it'd be worth adding a test in gporca.sql with the test case in the issue

But I found reproduce Eageragg plan need mass data, like `insert into foo select i%10000, i %1000, i from generate_series(1,100000000) i;` and I didn't other ways reproduce such  Eageragg plan. So could you give some brilliant  ideas to add such cases into gp_orca.sql.

**Comment by chrishajas** [comment]
> > Nice find! I also think it'd be worth adding a test in gporca.sql with the test case in the issue
> 
> But I found reproduce Eageragg plan need mass data, like `insert into foo select i%10000, i %1000, i from generate_series(1,100000000) i;` and I didn't other ways reproduce such Eageragg plan. So could you give some brilliant ideas to add such cases into gp_orca.sql.

Ah, yeah we don't want to add a lot of data. The following repros the issue on my machine with much less data (which I think is reasonable):

```
create table foo (j1 int, g1 int, s1 int);
insert into foo select i%10, i %10, i from generate_series(1,100) i;

create table bar (j2 int, g2 int, s2 int);
insert into bar select i%1, i %10, i from generate_series(1,10) i;

analyze foo;
analyze bar;

set optimizer = on;
set optimizer_enable_eageragg = on;

explain select max(s1) from foo inner join bar on j1 = j2 group by g1;
```

**Review by chrishajas — APPROVED** [review]
Looks good! Feel free to merge this

File: src/backend/gporca/libgpopt/src/xforms/CXformEagerAgg.cpp (hunk 2 above)
Comment (charliettxx) [reply src/backend/gporca/libgpopt/src/xforms/CXformEagerAgg.cpp]: done.

**Pull request merged**